lms/tasks/email_digests.py

```diff
@@ -90,12 +98,27 @@ def send_instructor_email_digests(
     updated_after = datetime.fromisoformat(updated_after)
     updated_before = datetime.fromisoformat(updated_before)
 
+    # How long to wait (in seconds) before retrying the task if it fails.
+    retry_countdown = (3600 * (self.request.retries + 1)) + random.randint(0, 900)
+
     with app.request_context() as request:  # pylint:disable=no-member
         with request.tm:
             digest_service = request.find_service(DigestService)
-            digest_service.send_instructor_email_digests(
-                h_userids,
-                updated_after,
-                updated_before,
-                override_to_email=override_to_email,
-            )
+
+            try:
+                digest_service.send_instructor_email_digests(
+                    h_userids,
+                    updated_after,
+                    updated_before,
+                    override_to_email=override_to_email,
+                )
+            except SendDigestsError as err:
+                self.retry(
+                    kwargs={
+                        **self.request.kwargs,
+                        "h_userids": list(err.errors.keys()),
+                    },
+                    countdown=retry_countdown,
+                )
+            except Exception:  # pylint:disable=broad-exception-caught
+                self.retry(countdown=retry_countdown)
```

Comment:
I've taken a fairly careful look at what exceptions this task might raise:

* I don't think the [`send_instructor_email_digests()` task's own code](https://github.com/hypothesis/lms/blob/64aaeb340f01b0871248467fb5435d526f755e98/lms/tasks/email_digests.py#L96-L124) is likely to raise any exceptions outside of the call to `DigestService.send_instructor_email_digests()` that's within the `try` block
* [`DigestService.send_instructor_email_digests()`](https://github.com/hypothesis/lms/blob/64aaeb340f01b0871248467fb5435d526f755e98/lms/services/digest.py#L50-L91) could raise an exception from the h API call that it makes or from the DB queries that it makes (via `DigestContext`) (for example if there's an issue with Postgres). These are the reason for the additional `except Exception` here: in the case of unexpected exceptions raised by `DigestService` we just want to retry the entire task.
* `DigestService` can deliberately raise `SendDigestsError`, which is handled specially by the `except SendDigestsError` above
* The only thing outside of `digest.py` that `DigestService` calls is [`MailchimpService.send_template()`](https://github.com/hypothesis/lms/blob/64aaeb340f01b0871248467fb5435d526f755e98/lms/services/mailchimp.py#L51-L77) and that doesn't look likely to raise anything unexpected, just the deliberately-raised `MailchimpError` which `DigestService` catches

Reply:
Hmm, I should probably add `LOG.exception()` and `h_pyramid_sentry.report_exception()` to the `except Exception` block. In the case of the `except SendDigestsError` block the `MailchimpError`'s will [already have been reported](https://github.com/hypothesis/lms/blob/64aaeb340f01b0871248467fb5435d526f755e98/lms/services/digest.py#L87-L88)

Reply:
Done